### PR TITLE
Pill Press

### PR DESCRIPTION
**Pull request by GenericDM**
## Pill Press!
This will add a lacking feature in ghettochem, the ability to produce pills.

### todo

- [x] Make it work!
- [x] Figure out how to select what pill type you want to use, when altclick, screwdriver, and wrench are all taken!
- [x] Learn how overlays work!
- [x] Sprites!
- [x] Get yelled at by coders probably!
- [x] Re-path it to be more in-line with other chem devices
- [x] Add it to cargo
---
As of now, you can insert a beaker, with optional pill bottle, to produce pills up to 30u.
You can adjust the size of the pills produced with a screwdriver.
The machine can be placed onto tables, or dragged around.

:cl:
add: Adds a hand-powered pill press for ghetto chemistry.
/:cl:

### CODE REVIEW

**Comment by GenericDM** [comment]
i should learn how to squash commits, that would be a good idea

**Review by Imaginos16 — APPROVED** [review]
Sprotes are fine for now, good hackjob honestly.

**Review by MarkSuckerberg — APPROVED** [review]
seems good!

File: code/modules/reagents/chemistry/machinery/chem_press.dm [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+		. += "<span class='notice'>The <b>pill bottle</b> slot is empty.</span>"
+	if(!beaker)
+		. += "<span class='notice>The <b>beaker</b> slot is empty.</span>"
+
+/obj/machinery/chem_press/attack_hand(mob/user)
+	. = ..()
+	if(.)
+		return
+	if(!beaker)
+		to_chat(user, "<span class='warning'>There's no container in [src]!.</span>")
```
Comment (thgvr) [inline code/modules/reagents/chemistry/machinery/chem_press.dm]:
```suggestion
		balloon_alert(user, "no container!")
```

File: code/modules/reagents/chemistry/machinery/chem_press.dm [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+		return FALSE
+	if(!beaker.reagents.total_volume)
+		to_chat(user, "<span class='warning'>[beaker] is empty!")
```
Comment (thgvr) [inline code/modules/reagents/chemistry/machinery/chem_press.dm]:
```suggestion
		balloon_alert(user, "[beaker] empty!")
```

File: code/modules/reagents/chemistry/machinery/chem_press.dm [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+		var/obj/item/reagent_containers/pill/P
+		// Check if there is a bottle that isn't full, then place a pill in the bottle. Otherwise, drop it on my tile.
+		if(bottle && bottle.contents.len < bottle.GetComponent(/datum/component/storage).max_items)
+			P = new/obj/item/reagent_containers/pill(bottle)
+		else
+			P = new/obj/item/reagent_containers/pill(drop_location())
+		P.name = "pill"
+		P.icon_state = "pill[pill_style]"
+		beaker.reagents.trans_to(P, current_volume, transfered_by = user)
+		to_chat(user, "<span class='notice'>You finish pressing a pill.</span>")
```
Comment (thgvr) [inline code/modules/reagents/chemistry/machinery/chem_press.dm]:
```suggestion
		balloon_alert(user, "finished pressing")
```

File: code/modules/reagents/chemistry/machinery/chem_press.dm [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	return TRUE
+
+/obj/machinery/chem_press/attackby(obj/item/I, mob/living/user, params)
```
Comment (thgvr) [inline code/modules/reagents/chemistry/machinery/chem_press.dm]: Descriptive vars, no `I` please

File: code/modules/reagents/chemistry/machinery/chem_press.dm [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	if(user.a_intent == INTENT_HARM) // Hit the machine if we're on harm intent.
+		return ..()
+	if(default_unfasten_wrench(user, I))
+		return
+	else if (istype(I, /obj/item/reagent_containers) && !(I.item_flags & ABSTRACT) && I.is_open_container())
+		. = TRUE //no afterattack
+		if(!user.transferItemToLoc(I, src))
+			return
+		handle_container(user, I)
+		to_chat(user, "<span class='notice'>You add [I] to the input slot [src].</span>")
```
Comment (thgvr) [inline code/modules/reagents/chemistry/machinery/chem_press.dm]:
```suggestion
		balloon_alert(user, "[I] added to input")
```

File: code/modules/reagents/chemistry/machinery/chem_press.dm [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+		return TRUE //no afterattack
+	else if(istype(I, /obj/item/storage/pill_bottle))
+		if(!user.transferItemToLoc(I, src))
+			return
+		handle_container(user, I)
+		to_chat(user, "<span class='notice'>You add [I] into the output slot of [src].</span>")
```
Comment (thgvr) [inline code/modules/reagents/chemistry/machinery/chem_press.dm]:
```suggestion
		balloon_alert(user, "[I] added to output")
```

File: code/modules/reagents/chemistry/machinery/chem_press.dm [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+		if(user.a_intent == INTENT_HELP)
+			var/i=0
+			for(var/A in possible_volumes)
+				i++
+				if(A == current_volume)
+					if(i<possible_volumes.len)
+						current_volume = possible_volumes[i+1]
+					else
+						current_volume = possible_volumes[1]
+					to_chat(user, "<span class='notice'>You adjust the press to produce [current_volume]u pills.</span>")
```
Comment (thgvr) [inline code/modules/reagents/chemistry/machinery/chem_press.dm]:
```suggestion
		balloon_alert(user, "making [current_volume]u pills")
```

File: code/modules/reagents/chemistry/machinery/chem_press.dm [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+		if(user.a_intent == INTENT_DISARM)
+			var/i=0
+			for(var/A in possible_styles)
+				i++
+				if(A == pill_style)
+					if(i<possible_styles.len)
+						pill_style = possible_styles[i+1]
+					else
+						pill_style = possible_styles[1]
+					to_chat(user, "<span class='notice'>You adjust the press to produce [style_colors["[pill_style]"]] pills.</span>")
```
Comment (thgvr) [inline code/modules/reagents/chemistry/machinery/chem_press.dm]:
```suggestion
		balloon_alert(user, "making [style_colors["[pill_style]"]] pills")
```

File: code/modules/reagents/chemistry/machinery/chem_press.dm [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+	if(beaker)
+		// If the container to handle is a beaker, try to swap it with the existing beaker.
+		if(istype(new_container, /obj/item/reagent_containers) || !new_container)
+			if(Adjacent(src, user) && !issiliconoradminghost(user))
+				user.put_in_hands(beaker)
+			else
+				beaker.forceMove(get_turf(src))
+			beaker = null
+			if(!new_container)
+				update_icon()
```
Comment (thgvr) [inline code/modules/reagents/chemistry/machinery/chem_press.dm]:
```suggestion
				update_appearance()
```

File: code/modules/reagents/chemistry/machinery/chem_press.dm [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+	if(bottle)
+		// If the container to handle is a bottle, try to swap it with the existing bottle.
+		if(istype(new_container, /obj/item/storage/pill_bottle) || !new_container)
+			if(Adjacent(src, user) && !issiliconoradminghost(user))
+				user.put_in_hands(bottle)
+			else
+				bottle.forceMove(get_turf(src))
+			bottle = null
+			if(!new_container)
+				update_icon()
```
Comment (thgvr) [inline code/modules/reagents/chemistry/machinery/chem_press.dm]:
```suggestion
				update_appearance()
```

File: code/modules/reagents/chemistry/machinery/chem_press.dm [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+				return TRUE
+	if(istype(new_container, /obj/item/storage/pill_bottle))
+		bottle = new_container
+	update_icon()
```
Comment (thgvr) [inline code/modules/reagents/chemistry/machinery/chem_press.dm]:
```suggestion
	update_appearance()
```

**Review by thgvr — CHANGES_REQUESTED** [review]

**Comment by thgvr** [comment]
Also, do you think you could add a construction recipe for this? It'd be useful even with outpost ordering. Stock part support optional.

Another thing - do not use both `to_chat` and `balloon_alert`, one or the other.

File: code/modules/reagents/chemistry/machinery/chem_press.dm [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+	if(.)
+		return
+	if(!beaker)
+		to_chat(user, "<span class='warning'>There's no container in [src]!.</span>")
+		return FALSE
+	if(!beaker.reagents.total_volume)
+		to_chat(user, "<span class='warning'>[beaker] is empty!")
+		return FALSE
+	if(do_after(user, press_time, target = src))
+		var/obj/item/reagent_containers/pill/P
```
Comment (thgvr) [inline code/modules/reagents/chemistry/machinery/chem_press.dm]: Descriptive var names

**Comment by GenericDM** [comment]
> Also, do you think you could add a construction recipe for this? It'd be useful even with outpost ordering. Stock part support optional.

What would you suggest for this? I had intended on making a recipe, but couldn't come up with any ideas for it.

**Review by thgvr — APPROVED** [review]

**Comment by Imaginos16** [comment]
im merging this lmao

**Pull request merged**